cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -264,43 +265,62 @@ fun dataFlowWithValidator(
-            setOf(this)
-        }
-        this.astParent is AssignExpression &&
-            this in (this.astParent as AssignExpression).rhs &&
-            (this.astParent as AssignExpression).operatorCode in
-                this.language.compoundAssignmentOperators -> {
-            // If we're the rhs of an assignment with an operator like +=, we should track the lhs
-            // value separately.
-            (this as? HasType)?.let { (this.astParent as AssignExpression).findTargets(it) }
-                ?: setOf()
-        }
-        this is Expression &&
+    val splitNodes =
+        when {
+            this is BinaryOperator && !this.type.isMutable -> {
+                // If the type of the node is a primitive (more precisely immutable) type, we will
+                // create new "object" that we have to track. But we should find the first
+                // declaration
+                // this flows to and track that one. If it flows to a reference first, this will
+                // typically be identified by handling the assignments below which is why we
+                // explicitly
+                // exclude such a flow here.
+                this.followDFGEdgesUntilHit(earlyTermination = { node, _ -> node is Reference }) {
+                        it is Declaration
+                    }
+                    .fulfilled
+                    .map { it.last() }
+                    .toSet()
+            }
+            /* A new object is constructed and our data flow into this object -> track the new object. */
+            this is ConstructExpression ||
+                /* A collection comprehension (e.g. list, set, dict comprehension) generates a new object similar to calling the constructor. */
+                this is CollectionComprehension -> {
+                setOf(this)
+            }
             this.astParent is AssignExpression &&
-            this in (this.astParent as AssignExpression).rhs &&
-            !this.type.isMutable -> {
-            (this.astParent as AssignExpression).findTargets(this)
+                this in (this.astParent as AssignExpression).rhs &&
+                (this.astParent as AssignExpression).operatorCode in
+                    this.language.compoundAssignmentOperators -> {
+                // If we're the rhs of an assignment with an operator like +=, we should track the
+                // lhs
+                // value separately.
+                (this as? HasType)?.let { (this.astParent as AssignExpression).findTargets(it) }
+                    ?: setOf()
+            }
+            this is Expression &&
+                this.astParent is AssignExpression &&
+                this in (this.astParent as AssignExpression).rhs &&
+                !this.type.isMutable -> {
+                (this.astParent as AssignExpression).findTargets(this)
+            }
+            else -> emptySet()
         }
-        else -> emptySet()
-    }
+
+    this.assume(
```

Comment:
Maybe the master of assumptions (@konradweiss) needs to have a word here too, but I would assume that we would split up this list that you have and only put the matching assumptions in the "when" branches? If that is even possible?

Reply:
Sounds plausible. However, for the `else` branch, I guess we need this whole text as we would have to validate that the current node did not miss any "DFG-splitting" in the behavior? +1 for a check by @konradweiss that we exploit all capabilities of the API and what is the anticipated usage